src/nvim/terminal.c

```diff
@@ -152,6 +154,111 @@ static VTermScreenCallbacks vterm_screen_callbacks = {
   .sb_popline  = term_sb_pop,
 };
 
+/// URL decoding (also know as Percent-encoding).
+///
+/// Note this function currently is only used for decoding shell's
+/// OSC 7 escape sequence which we can assume all bytes are valid
+/// UTF-8 bytes. Thus we don't need to deal with invalid UTF-8
+/// encoding bytes like 0xfe, 0xff.
+static size_t url_decode(const char *src, const size_t len, char_u *dst)
+{
+  size_t i = 0, j = 0;
+
+  while (i < len) {
+    if (src[i] == '%' && i + 2 < len) {
+      dst[j] = (char_u)hexhex2nr((char_u *)&src[i + 1]);
+      j++;
+      i += 3;
+    } else {
+      dst[j] = (char_u)src[i];
+      i++;
+      j++;
+    }
+  }
+  dst[j] = '\0';
+  return j;
+}
+
+/// Sync terminal buffer's cwd with shell's pwd with the help of OSC 7.
+///
+/// The OSC 7 sequence has the format of
+/// "\033]7;file://HOSTNAME/CURRENT/DIR\033\\"
+static void sync_shell_dir(const char *command, size_t cmdlen, void *user)
```

Comment:
this is a lot of new C code. could be avoided if we had a generic OSC event or Lua callback: 

- https://github.com/neovim/neovim/issues/4413

Reply:
Thanks for taking the time to review this PR!

>this is a lot of new C code. could be avoided (and more extensible) if we had a generic OSC event or Lua callback:

Yes, I kept thinking that making it a lua feature might be a better solution. I'll try to whip up a PR to help decide what the best approach is.